File: javascript/signature_changes.js
/**
 * JavaScript test file for function signature changes.
 * 
 * This file contains functions that will have their signatures modified in various ways
 * to test DiffScope's ability to detect signature-level changes.
 */

/**
 * This function will have parameters added in a later commit.
 * @param {number} a - First parameter
 * @param {number} b - Second parameter
 * @param {number} [c=0] - Third parameter (added)
 * @param {object} [d=null] - Fourth parameter (added)
 * @param {object} [options={}] - Options object (added)
 * @returns {number} Sum of parameters and number of options
 */
function functionForParamAddition(a, b, c = 0, d = null, options = {}) {
  let result = a + b;
  if (c) {
    result += c;
  }
  
  // Count the options
  const optionsCount = Object.keys(options).length;
  
  return result + optionsCount;
}

/**
 * This function will have parameters removed in a later commit.
 * @param {number} a - First parameter
 * @param {number} b - Second parameter
 * @returns {number} Sum of remaining parameters
 */
function functionForParamRemoval(a, b) {
  // The c and d parameters have been removed
  return a + b;
}

/**
 * A function designed to test parameter renames.
 * 
 * @param {number} first - First parameter (renamed from a)
 * @param {number} second - Second parameter (renamed from b)
 * @returns {number} Sum of the parameters
 */
function functionForParamRename(first, second) {
    return first + second;
}

/**
 * A function designed to test default parameter value changes.
 * 
 * @param {number} a - First parameter (no default)
 * @param {number} [b=20] - Second parameter (default changed from 10 to 20)
 * @param {number} [c=30] - Third parameter (default changed from 5 to 30)
 * @returns {number} Sum of all parameters
 */
function functionForDefaultParamChange(a, b = 20, c = 30) {
    return a + b + c;
}

/**
 * This function uses rest parameters and will be modified.
 * @param {string} prefix - Prefix to prepend to items
 * @param {...any} items - Items to process
 * @returns {Array<string>} Array of processed items
 */
function functionWithRestParams(prefix, ...items) {
  return items.map(item => `${prefix}:${item}`);
}

/**
 * This function uses destructuring in parameters and will be modified.
 * @param {Object} options - Options object
 * @param {string} options.name - Name property
 * @param {number} options.age - Age property
 * @returns {string} Formatted string with the options
 */
function functionWithDestructuring({ name, age }) {
  return `Name: ${name}, Age: ${age}`;
}

/**
 * This function will undergo multiple signature changes at once.
 * @param {number} a - First parameter
 * @param {number} b - Second parameter
 * @param {number} [c=null] - Optional third parameter
 * @returns {Object} Object containing parameter values
 */
function functionWithMultipleSignatureChanges(a, b, c = null) {
  const result = {
    a: a,
    b: b
  };
  
  if (c !== null) {
    result.c = c;
  }
  
  return result;
}

module.exports = {
  functionForParamAddition,
  functionForParamRemoval,
  functionForParamRename,
  functionForDefaultParamChange,
  functionWithRestParams,
  functionWithDestructuring,
  functionWithMultipleSignatureChanges
}; 

File: python/signature_changes.py
"""
Python test file for function signature changes.

This file contains functions that will have their signatures modified in various ways
to test DiffScope's ability to detect signature-level changes.
"""


def function_for_param_addition(a, b, c=0, d=None, **kwargs):
    """
    This function will have parameters added in a later commit.
    
    Args:
        a: First parameter
        b: Second parameter
        c: Third parameter (added)
        d: Fourth parameter (added)
        **kwargs: Additional keyword arguments (added)
        
    Returns:
        Sum of parameters and count of kwargs
    """
    result = a + b
    if c:
        result += c
    
    # Count the kwargs
    kwargs_count = len(kwargs)
    
    return result + kwargs_count


def function_for_param_removal(a, b):
    """
    This function will have parameters removed in a later commit.
    
    Args:
        a: First parameter
        b: Second parameter
        
    Returns:
        Sum of remaining parameters
    """
    # The c and d parameters have been removed
    return a + b


def function_for_param_rename(first, second):
    """
    A function designed to test parameter renames.
    
    Parameters:
    first (int): First parameter (renamed from a)
    second (int): Second parameter (renamed from b)
    
    Returns:
    int: Sum of the parameters
    """
    return first + second


def function_for_default_param_change(a, b=20, c=30):
    """
    A function designed to test default parameter value changes.
    
    Parameters:
    a (int): First parameter (no default)
    b (int): Second parameter (default changed from 10 to 20)
    c (int): Third parameter (default changed from 5 to 30)
    
    Returns:
    int: Sum of all parameters
    """
    return a + b + c


def function_with_type_annotations(a: int, b: str) -> str:
    """
    This function has type annotations that will be modified.
    
    Args:
        a: An integer
        b: A string
        
    Returns:
        A string with the integer repeated by specified amount
    """
    return b * a


def function_with_args_kwargs(*args, **kwargs):
    """
    This function uses *args and **kwargs and will be modified.
    
    Args:
        *args: Positional arguments
        **kwargs: Keyword arguments
        
    Returns:
        Dictionary containing args and kwargs
    """
    return {
        "args": args,
        "kwargs": kwargs
    }


def function_with_multiple_signature_changes(a, b, c=None):
    """
    This function will undergo multiple signature changes at once.
    
    Args:
        a: First parameter
        b: Second parameter
        c: Optional third parameter
        
    Returns:
        Dictionary containing parameter values
    """
    result = {
        "a": a,
        "b": b
    }
    
    if c is not None:
        result["c"] = c
        
    return result 